evm_loader/program/src/evm/database.rs

```diff
@@ -51,3 +49,43 @@ pub trait Database {
         is_static: bool,
     ) -> Option<Result<Vec<u8>>>;
 }
+
+/// Provides convenience methods that can be implemented in terms of `Database`.
+#[maybe_async(?Send)]
+pub trait DatabaseExt {
+    /// Returns whether an account exists, which is also referred to as the account being non-empty,
+    /// in accordance with https://eips.ethereum.org/EIPS/eip-161.
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool>;
+
+    /// Returns the code has for an address in accordance with https://eips.ethereum.org/EIPS/eip-1052.
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]>;
+}
+
+#[maybe_async(?Send)]
+impl<T: Database> DatabaseExt for T {
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool> {
+        Ok(self.nonce(address, chain_id).await? > 0 || self.balance(address, chain_id).await? > 0)
+    }
+
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]> {
+        // The function `Database::code` returns a zero-length buffer if the account exists with
+        // zero-length code, but also when the account does not exist. This makes it necessary to
+        // also check if the account exists when the returned buffer is empty.
+        //
+        // We could simplify the implementation by checking if the account exists first, but that
+        // would lead to more computation in what we think is the common case where the account
+        // exists and contains code.
+        let code = self.code(address).await?;
+        let bytes_to_hash = if !code.is_empty() {
+            Some(&*code)
+        } else if self.account_exists(address, chain_id).await? {
+            Some(<&[u8]>::default())
+        } else {
+            None
+        };
+
+        Ok(bytes_to_hash
+            .map(|bytes| solana_program::keccak::hash(bytes).to_bytes())
+            .unwrap_or_default())
```

Comment:
For me, it is hard to understand what the method does on the `Some(<&[u8]>::default())` flow  where `<&[u8]>::default()` actually means `&[]`or `None` flow  where `.unwrap_or_default()` is for the `None` in the if chain above.

Sometimes a little duplication can make the code more straightforward. I don't see the purpose of optional in this method. Why not just:
```Rust
        Ok(if !code.is_empty() {
            keccak::hash(&code).to_bytes()
        } else if self.account_exists(address, chain_id).await? {
            keccak::hash(&[]).to_bytes()
        } else {
            [0; 32]
        })
```
It is clear that we compute the hash of an empty slice for empty accounts and that we also return a zeroed array. I believe `Default` trait should not be used everywhere where it can be used if the code is less clear. This is low-level code and it needs to be as explicit as possible.

Low-level code also is easier to read and debug when using early returns:
```Rust
        if !code.is_empty() {
            return Ok(keccak::hash(&code).to_bytes());
        }

        if self.account_exists(address, chain_id).await? {
            return Ok(keccak::hash(&[]).to_bytes());
        }

        Ok([0; 32])
```

These are just suggestions, but the actual code seems a bit over-engineered in my opinion.